Ethereum trie proof verification

hsaleemsupra opened the pull request:
* Added verify_eth_trie_inclusion_proof and  verify_eth_trie_exclusion_proof methods to verify inclusion/exclusion proofs for ethereum trie
* Added native function to interact with the eth trie library to verify the proof
* Added tests 
* Added new gas constants for proof verification native function
* The max proof length limit is set to 2048 to prevent malicious calls to the function but it can be adjusted based on suggestions.
This link mentions the depth of the tree can be maximum 64: 
https://medium.com/coinmonks/5p-1r-ethereums-modified-merkle-patricia-trie-6956f5888398

So if the proof has a maximum of 16 nodes at each level proof size would be 1024. I selected 2048 to be safe.

hsaleemsupra committed:
adds eth_trie proof verification

hsaleemsupra committed:
fixes verify_proof, adds tests

hsaleemsupra committed:
update eth verify proof

* Instead of a single verify proof method, now we have verify inclusion/exclusion proof
* Updated gas cost

hsaleemsupra committed:
update max limit of proof nodes

hsaleemsupra committed:
minor change

hsaleemsupra committed:
adds rlp encoding/decoding for primitive types

isaacdoidge commented on aptos-move/aptos-gas-schedule/src/gas_schedule/aptos_framework.rs:
```diff
@@ -230,6 +230,15 @@ crate::gas_schedule::macros::define_gas_parameters!(
         [hash_keccak256_base: InternalGas, { 1.. => "hash.keccak256.base" }, 14704],
         [hash_keccak256_per_byte: InternalGasPerByte, { 1.. => "hash.keccak256.per_byte" }, 165],
 
+        [eth_trie_proof_base: InternalGas, { 12.. => "eth.trie.proof.base" }, 15000],
```
How did you choose the values for these parameters?

hsaleemsupra replied:
For now this is chosen to be somewhat higher than the base cost of keccak hashing. But any suggestions are welcome on what should we set this parameter to be.

isaacdoidge replied:
I meant for all of the new functions actually. We need to tune them according to the time that they take to execute.

hsaleemsupra replied:
Yes this needs to be done.

isaacdoidge commented on aptos-move/framework/aptos-stdlib/sources/rlp.move:
```diff
@@ -0,0 +1,359 @@
+    // 2) Encode/Decode for Primitives
+    //
+    //    Each function:
+    //      - BCS-serialize the primitive
```
Does RLP not have the ability to serialize the types directly?

hsaleemsupra replied:
RLP does have that functionality but that would involve creating the native encode/decode functions for each type. Let me see if we can have that instead.

isaacdoidge replied:
I assume that converting to BCS first will be more expensive, and if that is the case then it would be best to have native functions for each primitive.

hsaleemsupra replied:
I agree. The latest commit removes the bcs serialization and only uses native rlp encode/decode for each type

gorman-jacob commented on aptos-move/framework/src/natives/cryptography/eth_trie.rs:
```diff
@@ -0,0 +1,175 @@
+const HASHED_LENGTH: usize = 32;
+
+/// The maximum number of nodes allowed in a proof to prevent malicious calls to the function
+const MAX_PROOF_NODES: usize = 2048;
```
shouldn't MAX_PROOF_NODES be 64 since that is the max MPT depth? (unless I am misunderstanding how the proof is represented, or the key length is not 256-bits. `MAX_PROOF_NODES = 2048` should account for 1024-BYTE proof)

gorman-jacob replied:
EDIT: 65 since `value` is also included in the proof

gorman-jacob replied:
EDIT AGAIN: `proof` only includes `value` in the `eth_trie` implementation of `get_proof`. That is not necessarily true here. It would be helpful for several of my comments to know how `proof` is formatted

hsaleemsupra replied:
Yes you are right. I will make the change. Thanks.

gorman-jacob replied:
It would probably be prudent to add `const KEY_LENGTH: usize = 32;` and set `const MAX_PROOF_NODES = 2 * KEY_LENGTH (+1);`. We can then also check `key.len() == KEY_LENGTH` at the beginning of `native_verify_proof_eth_trie()`

hsaleemsupra replied:
The key.len() does not have to be 32 for example in this case:

https://github.com/Entropy-Foundation/aptos-core/blob/05348a39362a62f0f90c839a7b8554259b9c2321/aptos-move/framework/aptos-stdlib/sources/cryptography/eth_trie.move#L60

hsaleemsupra replied:
Also it seems like the eth_trie crate itself doesnot impose any restrictions on the key size.

hsaleemsupra replied:
We can set the limit on the key size and proof but it could be problematic in case we get a proof with key size > 32. The actual reason behind setting the max proof size was to restrict the caller from calling the method with a huge proof which may cause the verify method to slow down.

hsaleemsupra replied:
I will remove these limits for now. Until we are sure we would not be getting any keys/proofs larger than some limit.

gorman-jacob commented on aptos-move/framework/src/natives/cryptography/eth_trie.rs:
```diff
@@ -0,0 +1,175 @@
+///
+/// # Returns
+///
+/// A tuple of `(bool, vector<u8>)` where:
```
Documentation was confusing. "i.e. {exclusion/inclusion} proof" looked like it was describing the FOLLOWING case not the previously described case. Can be remedied by either 
1. Putting "i.e. ..." in parentheses and moving to previous line
2. Adding "i.e. invalid proof" after the final line so it is clear there is one "i.e." for each case
(I know this is nitpicky but I had to reread several times because I was confused)

hsaleemsupra replied:
Will address this.

gorman-jacob commented on aptos-move/framework/src/natives/cryptography/eth_trie.rs:
```diff
@@ -0,0 +1,175 @@
+/// A tuple of `(bool, vector<u8>)` where:
+///   - If the proof is valid and the key exists, returns `(true, value)` (with `value` being the found value)
+///     i.e. inclusion proof.
+///   - Otherwise, returns `(true, empty vector)` to show the proof is valid and key does not exist
```
Will the data associated with a key ever be an empty vector? If so this will cause ambiguity between exclusion proof and inclusion proof with no data

gorman-jacob replied:
Adding that this is totally possible within our code if allowed. `trie.get` on line 78 returns a wrapped Vec<u8> so if `value_opt == Some([])` then this proof of inclusion with empty data will be indistinguishable from an exclusion proof

hsaleemsupra replied:
Based on the eth_trie crate implementation, it is not possible to have an empty value corresponding to any key. So this situation should never arise. 

Ref:
https://github.com/carver/eth-trie.rs/blob/master/src/trie.rs#L313

gorman-jacob replied:
That check is only done for `EthTrie::insert(...)`. Here `trie` is initialized using `Trie::new(db)` and where `db` is initialized using `MemoryDB::insert(...)` which does not check if `value` is empty.

Ref:
https://github.com/carver/eth-trie.rs/blob/master/src/db.rs#L77
https://github.com/carver/eth-trie.rs/blob/master/src/db.rs#L58

hsaleemsupra replied:
The value can only be empty in db if the `node_encoded` is empty which is only possible if the `proof` has been tempered with after generating it. But in that case the root hash passed would not correspond to the new root hash. So the `trie.get` method would return an error. 
What do you think?

gorman-jacob commented on aptos-move/framework/src/natives/cryptography/eth_trie.rs:
```diff
@@ -0,0 +1,175 @@
+    for node_encoded in proof.iter() {
+        let hash: H256 = keccak(&node_encoded).as_fixed_bytes().into();
+        // Insert the node if it is the root or if its encoded length is at least HASHED_LENGTH.
+        if root_hash.eq(&hash) || node_encoded.len() >= HASHED_LENGTH {
```
I am a little confused by the condition `node_encoded.len() >= HASHED_LENGTH`. From my understanding, nodes with `node_encoded.len() <= HASHED_LENGTH` are stored inline in their parents. It is possible that I am misunderstanding how `proof` is formatted so I have broken it down into cases. Either:
1. `proof` contains a redundant copy of each of these "small nodes" which we are actively trying to exclude when building the EthTrie (in which case we should be checking `>` not `>=`)
2. `proof` does not contain redundant copies which means that either
    1. Small nodes are stored inline in their parents (which makes this check unnecessary)
    2. Small Nodes are *not* stored inline in their parents (which means that we are not adding those nodes to the tree in which case this is an issue which needs to be fixed)

Let me know if I am misunderstanding something.

hsaleemsupra replied:
This function is taken from the eth_trie crate for which we want to verify the merkle proofs. I am not sure about how the proof is formatted but this function is just a copy of the verify_proof method here:

https://github.com/carver/eth-trie.rs/blob/master/src/trie.rs#L463

gorman-jacob replied:
Probably has to do with the specifics of how they serialize nodes. My comment was mainly based on my high level understanding.

gorman-jacob commented on aptos-move/framework/src/natives/cryptography/eth_trie.rs:
```diff
@@ -0,0 +1,175 @@
+    let memdb = MemoryDB::new(true);
+    let db = Arc::new(memdb);
+    for node_encoded in proof.iter() {
+        let hash: H256 = keccak(&node_encoded).as_fixed_bytes().into();
```
hashing first is inefficient if `node_encoded.len() < HASHED_LENGTH`. If we assume `proof` is ordered than it is more efficient to check if this is the first iteration.

NOTE: this point is likely moot, see my comment about the `node_encoded.len() <= HASHED_LENGTH` condition

hsaleemsupra replied:
Again the function is taken from:

https://github.com/carver/eth-trie.rs/blob/master/src/trie.rs#L463

gorman-jacob commented on aptos-move/framework/src/natives/cryptography/eth_trie.rs:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    // Create an EthTrie instance using the temporary DB and the given root.
+    let trie = EthTrie::new(db).at_root(root_hash);
```
does `at_root` throw some kind of error if hash of the root stored in `db` does not match `root_hash`? If not, it is possible that `root_hash` does not match the hash of the root node (this can happen if the root node has `node_encoded.len() >= HASHED_LENGTH` in which case it bypasses the condition `root_hash.eq(&hash)`

hsaleemsupra replied:
`at_root` method does not throw any errors. However based on [at_root](https://github.com/carver/eth-trie.rs/blob/94ad815505c4a1dce97d6f30a052446ce3b2abfb/src/trie.rs#L238) method both `root` and `root_hash` are set in this method using the passed `root_hash` so they should be same.

If the db is invalid i.e. it has a wrong root node (or any other node), the get method below should give an error based on these calls:

https://github.com/carver/eth-trie.rs/blob/94ad815505c4a1dce97d6f30a052446ce3b2abfb/src/trie.rs#L398
https://github.com/carver/eth-trie.rs/blob/94ad815505c4a1dce97d6f30a052446ce3b2abfb/src/trie.rs#L259
https://github.com/carver/eth-trie.rs/blob/94ad815505c4a1dce97d6f30a052446ce3b2abfb/src/trie.rs#L446

gorman-jacob replied:
Spent a while tracing through the code and I think I follow all the checks that are happening. It's copied code anyways so should be fine.

hsaleemsupra committed:
updates to incorporate suggestions

- Native rlp encoding for various types instead of using bcs
- Removed Max proof size limit from eth trie

isaacdoidge reviewed (approved):
I'm approving this PR on the basis that the code will initially be deployed to a devnet, so the gas costs of the native functions can be tuned and updated in another PR. Same with their implementation so as to avoid the BCS serde.

Please wait for @gorman-jacob's approval before merging.
